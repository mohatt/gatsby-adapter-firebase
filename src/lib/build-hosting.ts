import type { RoutesManifest } from 'gatsby'
import {
  HeaderKV,
  FirebaseHostingHeader,
  FirebaseHostingRedirect,
  FirebaseHostingRewrite,
  FirebaseHostingFunctionRewrite,
  FunctionVariants,
} from './types.js'
import type { FunctionConfig } from './runtime/types.js'
import type { AdaptorReporter } from './reporter.js'

export interface BuildHostingArgs {
  routesManifest: RoutesManifest
  pathPrefix: string
  reporter: AdaptorReporter
  functionsMap: ReadonlyMap<string, FunctionVariants>
}

export interface BuildHostingResult {
  headers: FirebaseHostingHeader[]
  redirects: FirebaseHostingRedirect[]
  rewrites: FirebaseHostingRewrite[]
}

const REDIRECT_STATUS_CODES = [301, 302, 303, 307, 308]

const ensureLeadingSlash = (value: string) => {
  if (!value) return '/'
  return value.startsWith('/') ? value : `/${value}`
}

const splitLocation = (value: string) => {
  if (!value) return { path: '', suffix: '' }
  const index = value.search(/[?#]/)
  if (index === -1) return { path: value, suffix: '' }
  return {
    path: value.slice(0, index),
    suffix: value.slice(index),
  }
}

const applyPathPrefix = (value: string, pathPrefix: string) => {
  if (!pathPrefix) return value
  const normalizedPrefix = ensureLeadingSlash(pathPrefix).replace(/\/$/, '')
  if (!normalizedPrefix || normalizedPrefix === '/') {
    return value
  }

  if (value === '/') {
    return normalizedPrefix
  }

  if (value.startsWith(`${normalizedPrefix}/`) || value === normalizedPrefix) {
    return value
  }

  // collapse slashes
  return `${normalizedPrefix}${value}`.replace(/\/{2,}/g, '/')
}

const normalizeSource = (value: string, pathPrefix: string) => {
  const base = ensureLeadingSlash(value || '/')
  const prefixed = applyPathPrefix(base, pathPrefix)
  return prefixed.replace(/\/\*$/u, '/**') || '/'
}

const normalizeDestination = (value: string, pathPrefix: string) => {
  const { path, suffix } = splitLocation(value)
  const normalizedPath = applyPathPrefix(ensureLeadingSlash(path || '/'), pathPrefix) || '/'
  const collapsed = normalizedPath.replace(/\/{2,}/g, '/')
  return `${collapsed}${suffix ?? ''}`
}

const extractRegion = (options?: FunctionConfig): string | null => {
  if (!options) return null
  const region = options?.region
  if (typeof region === 'string') return region
  if (Array.isArray(region)) return region[0]
  if (region && 'value' in region && typeof region.value === 'function') return region.value()
  return null
}

export const buildHosting = (args: BuildHostingArgs): BuildHostingResult => {
  const { routesManifest, pathPrefix, reporter, functionsMap } = args
  const headerAccumulator = new Map<string, Map<string, HeaderKV>>()
  const redirects: FirebaseHostingRedirect[] = []
  const rewrites: FirebaseHostingRewrite[] = []

  const addHeaders = (source: string, entries: HeaderKV[]) => {
    if (!entries.length) return
    let bucket = headerAccumulator.get(source)
    if (!bucket) {
      bucket = new Map()
      headerAccumulator.set(source, bucket)
    }
    for (const { key, value } of entries) {
      bucket.set(key.toLowerCase(), { key, value })
    }
  }

  for (const route of routesManifest) {
    const { path: routePath, suffix: routeSuffix } = splitLocation(route.path)
    const source = normalizeSource(routePath, pathPrefix)

    if (route.type === 'function') {
      const variants = functionsMap.get(route.functionId)
      if (!variants) {
        reporter.warn(
          `Function route ${route.path} -> "${route.functionId}" has no matching function definition; skipping rewrite for ${source}`,
        )
        continue
      }
      if (route.cache && !variants.cached) {
        reporter.warn(
          `Function route ${route.path} -> "${route.functionId}" is marked cache=true but cached variant could not be generated; using default deployment.`,
        )
      }
      const functionEntry = route.cache && variants.cached ? variants.cached : variants.default
      const { deployId, config } = functionEntry
      const region = extractRegion(config)

      const rewrite: FirebaseHostingFunctionRewrite = {
        source,
        function: {
          functionId: deployId,
          pinTag: true,
          ...(region ? { region } : {}),
        },
      }
      rewrites.push(rewrite)
      continue
    }

    if (route.type === 'static') {
      addHeaders(source, route.headers)
      continue
    }

    if (route.type === 'redirect') {
      if (routeSuffix) {
        reporter.warn(
          `Redirect "${route.path}" -> "${route.toPath}" contains query parameters or hash fragments which Firebase Hosting cannot match; skipping this rule.`,
        )
        continue
      }

      if (route.headers?.length) {
        reporter.warn(
          `Redirect ${route.path} -> ${route.toPath} defines HTTP headers but Firebase Hosting redirects do not support response headers; omitting headers.`,
        )
      }
      if (route.ignoreCase) {
        reporter.warn(
          `Redirect ${route.path} -> ${route.toPath} sets ignoreCase=true which is not supported by Firebase Hosting; proceeding with case-sensitive match.`,
        )
      }
      const conditions = route['conditions'] as Record<string, unknown> | undefined
      if (conditions && Object.keys(conditions).length > 0) {
        reporter.warn(
          `Redirect ${route.path} -> ${route.toPath} has conditions (${Object.keys(conditions).join(', ')}) which are not supported by Firebase Hosting; skipping rewrite for ${source}`,
        )
        continue
      }

      const isExternal = /^https?:\/\//i.test(route.toPath)
      const destination = isExternal ? route.toPath : normalizeDestination(route.toPath, pathPrefix)

      if (route.status === 200) {
        if (isExternal) {
          reporter.warn(
            `Rewrite ${route.path} -> ${route.toPath} targets an external URL; Firebase Hosting rewrites cannot proxy to external origins. Falling back to 302 redirect.`,
          )
          redirects.push({ source, destination: route.toPath, type: 302 })
        } else {
          rewrites.push({ source, destination })
        }
        continue
      }

      if (!REDIRECT_STATUS_CODES.includes(route.status)) {
        reporter.warn(
          `Redirect ${route.path} -> ${route.toPath} uses unsupported status ${route.status}; skipping.`,
        )
        continue
      }

      redirects.push({
        source,
        destination,
        type: route.status,
      })
    }
  }

  // map to headers array
  const headers: FirebaseHostingHeader[] = Array.from(headerAccumulator.entries())
    .map(([source, headerMap]) => ({ source, headers: Array.from(headerMap.values()) }))
    .sort((a, b) => a.source.localeCompare(b.source))

  return { headers, redirects, rewrites }
}
