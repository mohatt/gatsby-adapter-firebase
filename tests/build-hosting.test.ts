import { createTestArgs } from '../test/util.js'
import { buildHosting, BuildHostingArgs } from '../src/lib/build-hosting.js'

describe('buildHosting()', () => {
  it('converts Gatsby routes into Firebase hosting rules', () => {
    const args = createTestArgs<BuildHostingArgs>({
      routesManifest: [
        { type: 'function', path: '/ssr', functionId: 'ssr-engine' },
        { type: 'function', path: '/ssr/page-data.json', functionId: 'ssr-engine' },
        { type: 'function', path: '/ssr-deferred', functionId: 'ssr-engine', cache: true },
        { type: 'redirect', path: '/docs/*', toPath: '/docs/index', status: 200, headers: [] },
        { type: 'redirect', path: '/en/docs/*', toPath: '/docs/*', status: 301, headers: [] },
        {
          type: 'redirect',
          path: '/old',
          toPath: '/new',
          status: 302,
          headers: [{ key: 'test', value: 'ignored' }],
        },
        {
          type: 'redirect',
          path: '/ext',
          toPath: 'https://www.awesomesite.com',
          status: 200,
          headers: [],
        },
        {
          type: 'redirect',
          path: '/legacy?tag=:id',
          toPath: '/blog/:id',
          status: 301,
          headers: [],
        },
        {
          type: 'static',
          path: '/static',
          filePath: 'public/static.html',
          headers: [{ key: 'cache-control', value: 'public, max-age=0' }],
        },
      ],
      functionsMap: new Map([
        [
          'ssr-engine',
          {
            default: { deployId: 'ssr_engine', config: {} },
            cached: { deployId: 'ssr_engine_cached', config: {} },
          },
        ],
      ]),
      pathPrefix: '',
    })
    const { config } = buildHosting(args)
    expect(config).toMatchSnapshot('config')
    expect(args.gatsbyReporter.warn.mock.calls).toMatchSnapshot('warnings')
  })

  it('converts redirect splats and params into Firebase-compatible patterns', () => {
    const args = createTestArgs<BuildHostingArgs>({
      routesManifest: [
        { type: 'redirect', path: '/old/*', toPath: '/new', status: 301, headers: [] },
        { type: 'redirect', path: '/old/*', toPath: '/new/*', status: 302, headers: [] },
        { type: 'redirect', path: '/old/path', toPath: '/new/path', status: 301, headers: [] },
        {
          type: 'redirect',
          path: '/old/:id/profile',
          toPath: '/new/:id/profile',
          status: 301,
          headers: [],
        },
        {
          type: 'redirect',
          path: '/old/:id/posts/*',
          toPath: '/new/:id/posts/*',
          status: 302,
          headers: [],
        },
        {
          type: 'redirect',
          path: '/old/*/post/*',
          toPath: '/new/*/article/*',
          status: 301,
          headers: [],
        },
        {
          type: 'redirect',
          path: '/old/*',
          toPath: '/new/*/page?foo=bar#head',
          status: 301,
          headers: [],
        },
        {
          type: 'redirect',
          path: '/old/*',
          toPath: '/new/*/?foo=bar#head',
          status: 301,
          headers: [],
        },
        {
          type: 'redirect',
          path: '/awesome/*',
          toPath: 'https://www.awesomesite.com/docs/*',
          status: 301,
          headers: [],
        },
        {
          type: 'redirect',
          path: '/awesome/*',
          toPath: 'https://www.awesomesite.com/docs/*/page?foo=bar#head',
          status: 301,
          headers: [],
        },
        {
          type: 'redirect',
          path: '/awesome/*/post/*',
          toPath: 'https://www.awesomesite.com/docs/*/article/*/?from=bar#head',
          status: 301,
          headers: [],
        },
      ],
      pathPrefix: '',
    })
    const { config } = buildHosting(args)
    expect(config).toMatchSnapshot('config')
    expect(args.gatsbyReporter.warn).not.toBeCalled()
  })

  it('derives rewrite region from config overrides with fallback to defaults', () => {
    const args = createTestArgs<BuildHostingArgs>({
      routesManifest: [
        { type: 'function', path: '/ssr', functionId: 'ssr-engine' },
        { type: 'function', path: '/api', functionId: 'hello-world' },
      ],
      functionsMap: new Map([
        [
          'ssr-engine',
          {
            default: { deployId: 'ssr_engine', config: { region: 'asia-northeast1' } },
            cached: { deployId: 'ssr_engine_cached', config: { region: 'europe-west1' } },
          },
        ],
        [
          'hello-world',
          { default: { deployId: 'hello_world', config: { region: 'europe-west1' } } },
        ],
      ]),
      pathPrefix: '',
    })
    const { config } = buildHosting(args)
    expect(config).toMatchSnapshot('config')
    expect(args.gatsbyReporter.warn).not.toBeCalled()
  })

  it('warns when cached route has no cached variant', () => {
    const args = createTestArgs<BuildHostingArgs>({
      routesManifest: [{ type: 'function', path: '/dsg', functionId: 'ssr-engine', cache: true }],
      functionsMap: new Map([['ssr-engine', { default: { deployId: 'ssr_engine', config: {} } }]]),
      pathPrefix: '',
    })
    const { config } = buildHosting(args)
    expect(config).toMatchSnapshot('config')
    expect(args.gatsbyReporter.warn.mock.calls).toMatchSnapshot('warnings')
  })

  it('applies pathPrefix only to internal destinations', () => {
    const args = createTestArgs<BuildHostingArgs>({
      pathPrefix: '/stage',
      routesManifest: [
        { type: 'redirect', path: '/docs/*', toPath: '/help/*', status: 301, headers: [] },
        { type: 'redirect', path: '/docs/static', toPath: '/landing', status: 302, headers: [] },
        {
          type: 'redirect',
          path: '/docs/*',
          toPath: 'https://external.example.com/docs/*',
          status: 301,
          headers: [],
        },
        {
          type: 'redirect',
          path: '/docs/*',
          toPath: 'https://external.example.com/docs/*/page?foo=bar#head',
          status: 301,
          headers: [],
        },
      ],
    })
    const { config } = buildHosting(args)
    expect(config).toMatchSnapshot('config')
    expect(args.gatsbyReporter.warn).not.toBeCalled()
  })
})
