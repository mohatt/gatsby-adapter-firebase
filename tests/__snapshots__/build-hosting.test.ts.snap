// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildHosting() > applies pathPrefix only to internal destinations > config 1`] = `
{
  "headers": [],
  "public": "public",
  "redirects": [
    {
      "destination": "/stage/help/:splat",
      "source": "/stage/docs/:splat*",
      "type": 301,
    },
    {
      "destination": "/stage/landing",
      "source": "/stage/docs/static",
      "type": 302,
    },
    {
      "destination": "https://external.example.com/docs/:splat",
      "source": "/stage/docs/:splat*",
      "type": 301,
    },
    {
      "destination": "https://external.example.com/docs/:splat/page?foo=bar#head",
      "source": "/stage/docs/:splat*",
      "type": 301,
    },
  ],
  "rewrites": [],
  "target": "test-functions",
}
`;

exports[`buildHosting() > converts Gatsby routes into Firebase hosting rules > config 1`] = `
{
  "headers": [
    {
      "headers": [
        {
          "key": "cache-control",
          "value": "public, max-age=0",
        },
      ],
      "source": "/static",
    },
  ],
  "public": "public",
  "redirects": [
    {
      "destination": "/docs/:splat",
      "source": "/en/docs/:splat*",
      "type": 301,
    },
    {
      "destination": "/new",
      "source": "/old",
      "type": 302,
    },
    {
      "destination": "https://www.awesomesite.com/",
      "source": "/ext",
      "type": 302,
    },
  ],
  "rewrites": [
    {
      "function": {
        "functionId": "ssr_engine",
      },
      "regex": "^/ssr(?:/)?$",
    },
    {
      "function": {
        "functionId": "ssr_engine",
      },
      "source": "/ssr/page-data.json",
    },
    {
      "function": {
        "functionId": "ssr_engine_cached",
      },
      "regex": "^/ssr-deferred(?:/)?$",
    },
    {
      "destination": "/docs/index",
      "source": "/docs/:splat*",
    },
  ],
  "target": "test-functions",
}
`;

exports[`buildHosting() > converts Gatsby routes into Firebase hosting rules > warnings 1`] = `
[
  [
    "[gatsby-adaptor-firebase]: Redirect /old -> /new defines HTTP headers but Firebase Hosting redirects do not support response headers; omitting headers.",
  ],
  [
    "[gatsby-adaptor-firebase]: Rewrite /ext -> https://www.awesomesite.com targets an external URL; Firebase Hosting rewrites cannot proxy to external origins. Falling back to 302 redirect.",
  ],
  [
    "[gatsby-adaptor-firebase]: Redirect "/legacy?tag=:id" -> "/blog/:id" contains query parameters or hash fragments which Firebase Hosting cannot match; skipping this rule.",
  ],
]
`;

exports[`buildHosting() > converts redirect splats and params into Firebase-compatible patterns > config 1`] = `
{
  "headers": [],
  "public": "public",
  "redirects": [
    {
      "destination": "/new",
      "source": "/old/:splat*",
      "type": 301,
    },
    {
      "destination": "/new/:splat",
      "source": "/old/:splat*",
      "type": 302,
    },
    {
      "destination": "/new/path",
      "source": "/old/path",
      "type": 301,
    },
    {
      "destination": "/new/:id/profile",
      "source": "/old/:id/profile",
      "type": 301,
    },
    {
      "destination": "/new/:id/posts/:splat",
      "source": "/old/:id/posts/:splat*",
      "type": 302,
    },
    {
      "destination": "/new/:splat/article/:splat1",
      "source": "/old/:splat*/post/:splat1*",
      "type": 301,
    },
    {
      "destination": "/new/:splat/page?foo=bar#head",
      "source": "/old/:splat*",
      "type": 301,
    },
    {
      "destination": "/new/:splat/?foo=bar#head",
      "source": "/old/:splat*",
      "type": 301,
    },
    {
      "destination": "https://www.awesomesite.com/docs/:splat",
      "source": "/awesome/:splat*",
      "type": 301,
    },
    {
      "destination": "https://www.awesomesite.com/docs/:splat/page?foo=bar#head",
      "source": "/awesome/:splat*",
      "type": 301,
    },
    {
      "destination": "https://www.awesomesite.com/docs/:splat/article/:splat1/?from=bar#head",
      "source": "/awesome/:splat*/post/:splat1*",
      "type": 301,
    },
  ],
  "rewrites": [],
  "target": "test-functions",
}
`;

exports[`buildHosting() > derives rewrite region from config overrides with fallback to defaults > config 1`] = `
{
  "headers": [],
  "public": "public",
  "redirects": [],
  "rewrites": [
    {
      "function": {
        "functionId": "ssr_engine",
        "region": "asia-northeast1",
      },
      "regex": "^/ssr(?:/)?$",
    },
    {
      "function": {
        "functionId": "hello_world",
        "region": "europe-west1",
      },
      "regex": "^/api(?:/)?$",
    },
  ],
  "target": "test-functions",
}
`;

exports[`buildHosting() > warns when cached route has no cached variant > config 1`] = `
{
  "headers": [],
  "public": "public",
  "redirects": [],
  "rewrites": [
    {
      "function": {
        "functionId": "ssr_engine",
      },
      "regex": "^/dsg(?:/)?$",
    },
  ],
  "target": "test-functions",
}
`;

exports[`buildHosting() > warns when cached route has no cached variant > warnings 1`] = `
[
  [
    "[gatsby-adaptor-firebase]: Function route /dsg -> "ssr-engine" is marked cache=true but cached variant could not be generated; using default deployment.",
  ],
]
`;
